### PR TITLE
Pallet dispatchable+storage doc module. 

### PR DESCRIPTION
**Pull request by kianenigma**
hopefully closes https://github.com/paritytech/substrate/issues/12398, but needs more work. 
step toward https://github.com/paritytech/substrate/issues/13299

## Description

The goal of this PR is to make docs for dispatchables and storages on pallets more accessible and discoverable when browsing the generated docs. The PR accomplishes this by adding two auto-generated modules to all pallets: `dispatchables` and `storage_types`. These modules are only produced when building docs and otherwise do not exist or interfere with pallets. The `storage_types` module is auto-populated with documented structs representing each storage type from the parent pallet. The `dispatchables` module is auto-populated with uncallable auto-generated functions matching the signature of each `Call` variant. Right now a link to the corresponding `Call` enum variant is also included, though this may be removed. A note is included specifying that these cannot be called and are doc-only.

Modules added to `pallet` module:

image

Detail view of `storage_types` module:

image

Detail view of of `dispatchables` module:

Screenshot 2023-03-14 at 5 11 58 PM

Detail view of a particular dispatchable:

Screenshot 2023-03-14 at 5 12 45 PM

Also hides docs for `Instance2`-`Instance16` on the `pallet` module and adds a doc comment to `Instance1` explaining this:

image

### CODE REVIEW

**Comment by sam0x17** [comment]
Figuring out how we want to handle generics in fn signatures in pallets aka situations like this:

```rust
pub fn set(origin: OriginFor<T>, #[pallet::compact] now: T::Moment) -> DispatchResult {
```
are causing this error:
```
error[E0412]: cannot find type `T` in this scope
   --> frame/balances/src/lib.rs:372:28
    |
369 |         pub fn force_transfer(
    |                              - help: you might be missing a type parameter: `<T>`
...
372 |             dest: AccountIdLookupOf<T>,
    |                                     ^ not found in this scope
```

As far as I can tell that is the only outstanding thing not working

**Comment by kianenigma** [comment]
> Figuring out how we want to handle generics in fn signatures in pallets aka situations like this:
> 
> ```rust
> 		#[pallet::call_index(2)]
> 		#[pallet::weight(T::WeightInfo::force_transfer())]
> 		pub fn force_transfer(
> 			origin: OriginFor<T>,
> 			source: AccountIdLookupOf<T>,
> 			dest: AccountIdLookupOf<T>,
> 			#[pallet::compact] value: T::Balance,
> 		) -> DispatchResultWithPostInfo {
> ```
> 
> are causing this error:
> 
> ```
> error[E0412]: cannot find type `T` in this scope
>    --> frame/balances/src/lib.rs:372:28
>     |
> 369 |         pub fn force_transfer(
>     |                              - help: you might be missing a type parameter: `<T>`
> ...
> 372 |             dest: AccountIdLookupOf<T>,
>     |                                     ^ not found in this scope
> ```
> 
> As far as I can tell that is the only outstanding thing not working

Maybe put a `<T: Config>` on every function? or better, wrap them all in 

```
/// Some doc saying that this struct does not exist in reality
pub struct Dispatchables<T: Config>(Phantom(T));
impl<T: Config> Dispatchables {
  // This is now 100% identical to `Pallet` struct and should work. 
}
```

and then we export `struct Dispatchables`, which is good enough.

**Comment by bkchr** [comment]
This clearly requires some screenshots on how that looks in the docs. I'm not 100% sure that adding fake items to the docs is such a great idea.

**Comment by sam0x17** [comment]
Screenshot 2023-03-09 at 12 17 51 AM

These when clicked take you to the relevant docs. I think the best dev UX would be to have the docs here verbatim but alas that doesn't quite work in rustdoc with re-exports. I actually ran into a literal compiler bug when I tried to do anything other than re-exporting for this one, btw.
Screenshot 2023-03-09 at 12 25 06 AM

This one works better because they are not re-exports:
Screenshot 2023-03-09 at 4 02 14 PM

**Comment by bkchr** [comment]
Did you tried using `pub type StorageItem = super::StorageItem` instead of the re-export.

But with the calls people will then not know how to use them. From the docs it looks like there is some module, which doesn't really exists 🤷

**Comment by sam0x17** [comment]
> Did you tried using `pub type StorageItem = super::StorageItem` instead of the re-export.
> 
> But with the calls people will then not know how to use them. From the docs it looks like there is some module, which doesn't really exists 🤷

Yeah in that case I think the current solution is the best compromise because at least it points them to the right place.

As far as accessibility, I will add a comment saying these types and functions are available on the pallet itself or something to that effect

**Comment by kianenigma** [comment]
Suggestion: For each dispatchable, link to the actually function in `struct Pallet` and the corresponding `enum Call` variant as well, emphasizing that "the real thing is there, this is just a fake" list. 

For storage items, I think this is great and already a massive improvement. 

I will try and help a bit with the auto-generated documents, making sure they are not confusing, especially for the dispatch stuff.

**Review by kianenigma — COMMENTED** [review]
I think it is a good step forward, but perhaps we can chose the doc strings a bit better. 

Can't approve as I opened the PR, but has my blessing.

**Comment by sam0x17** [comment]
Hmm I wonder if we could instead just add the docs to the Call enum variants? thoughts?

**Comment by KiChjang** [comment]
As with everything doc-related, it's hard to give comments if we don't know how it looks like on the UI, especially when it is about grouping similar items together, which is what the entire underlying issue is about.

**Comment by sam0x17** [comment]
> As with everything doc-related, it's hard to give comments if we don't know how it looks like on the UI, especially when it is about grouping similar items together, which is what the entire underlying issue is about.

i.e. what if we emit doc strings on these variants instead of the `dispatchables` module shown above:
Screenshot 2023-03-10 at 2 12 41 AM

**Comment by sam0x17** [comment]
lol so we've come full circle. I scrolled down the page on the `Call` enum only to realize the variants are _already_ documented:
Screenshot 2023-03-10 at 8 19 14 AM

Do we maybe just want to surface this better with a link to the `Call` variant on the main docs page for each pallet? Looks like this already perfectly documents the dispatchables for any given pallet

**Comment by sam0x17** [comment]
I guess the intent though of what @kianenigma was doing with the dispatchables was to render them in the docs as actual functions instead of enum variants. Given that, I think it might be better to leave it the way he set it up, but I'm going to add doc links to the actual variants like he suggested on each one

**Comment by sam0x17** [comment]
OK so yeah latest commit adds this preface to the docs for the auto-generated dispatchables stub functions:

Screenshot 2023-03-11 at 2 23 27 AM

Notably, the NOTE actually links to the proper enum variant corresponding with the function.

**Comment by sam0x17** [comment]
If we like the wording above I think this is ready

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522151

**Comment by sam0x17** [comment]
Have also added a revamp of the very ugly `Instance1`, `Instance2`, `Instance3`... `Instance16` stuff in the `pallet` module based on @ggwpez's feedback:

image

**Comment by ggwpez** [comment]
Using structs for the storage types works. We dont actually need the *concrete* storage type. If we really want we could still link to the `frame-support::storage::...` type but for high-level this looks good to me.
Before:
image
After:
image

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522449

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522443

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522442

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522444

**Review by ggwpez — APPROVED** [review]
Looks much better now, thanks!

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522506

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522499

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522501

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522500

**Comment by sam0x17** [comment]
@KiChjang this should be good for a final look now, screenshots at the top are updated

File: frame/support/procedural/src/pallet/expand/doc_only.rs
Comment (bkchr) [inline frame/support/procedural/src/pallet/expand/doc_only.rs:40]:
```suggestion
						" designed to document the [`{}`][`Call::{}`] variant of",
```

There is no requirement that `Call` is at the top of the crate.

File: frame/support/procedural/src/pallet/expand/doc_only.rs
Comment (bkchr) [inline frame/support/procedural/src/pallet/expand/doc_only.rs:50]:
```suggestion
						/// the pallet [`Call`](`Call`) enum. You should not attempt to
```

Same

File: frame/support/procedural/src/pallet/expand/doc_only.rs
Comment (bkchr) [inline frame/support/procedural/src/pallet/expand/doc_only.rs:48]:
```suggestion
						/// NOTE: This function is an automatically generated, doc only, uncallable stub.
```

File: frame/support/procedural/src/pallet/expand/doc_only.rs
Comment (bkchr) [inline frame/support/procedural/src/pallet/expand/doc_only.rs:24]:
```suggestion
	let storage_names = def.storages.iter().map(|storage| &storage.ident);
	let storage_docs = def.storages.iter().map(|storage| &storage.docs);
```

File: frame/support/procedural/src/pallet/expand/doc_only.rs
Comment (bkchr) [inline frame/support/procedural/src/pallet/expand/doc_only.rs:25]:
```suggestion
	let dispatchables =
```

File: frame/support/procedural/src/pallet/expand/doc_only.rs
Comment (bkchr) [inline frame/support/procedural/src/pallet/expand/doc_only.rs:59]:
```suggestion
		};
```

**Review by bkchr — APPROVED** [review]

**Review by KiChjang — APPROVED** [review]

**Comment by sam0x17** [comment]
bot rebase

**Comment by sam0x17** [comment]
bot merge

**Pull request merged**
